### DISCUSSION TITLE
[API Request] Clarification / Definition of covertArt handling when missing

### DISCUSSION THREAD

**Discussion by Tolriq** [opensubsonic/open-subsonic-api#39] [API requests]
### Type of change

API Clarification

### Proposal description

This is something to be decided as not currently documented.

1) Should servers be allowed to returned covertArt values even when there's no actual cover Art?
2) Should servers be allowed to return a default image when there's no actual images?

If all servers currently do that, then an extension should be created to add a parameter to getCovertArt to prevent returning default image and return a proper error 404 for example.

### Backward compatibility

- [ ] No backward compatibility impact.

### API details

TBD

**Comment by epoupon** [comment]
I am for not sending any default cover as the client should display its own placeholder.
We can use return code 70 (The requested data was not found) if the cover id does not exist or not resolvable for server that put artist ids or whatever in it?

**Reply by Tolriq** [reply]
The problem is the the function is usually called by actual image loader / renderers that have no idea what subsonic API and how to deal with json/xml in a 200 result that should have contained an image. Some may try to decode / cache the error, some may crash.
It's better to actually rely on normal HTTP error handling for this case (As it should be for stream / download too but the API documentation state otherwise so while it should be changed at some point this would be a breaking change)

**Reply by epoupon** [reply]
Ah yes forgot about that. A 404 should do the job indeed

**Comment by spl0k** [comment]
I fail to understand the difference between your two points.

Using some RFC-style wording:
If the server doesn't have any art for an entity, the `covertArt` field MUST be absent from this entity response representation. In the event the client receives an entity where this field is missing, it may display a default cover art of its choosing.
If the server receives a request on its `getCoverArt` endpoint with an `id` parameter value that doesn't match any art know by the server, it MUST respond with a 404 HTTP status code. The response body SHOULD include a Subsonic response with a 70 error code.

**Reply by Tolriq** [reply]
I forget the ? at the end those where questions due to how servers vastly vary on this particular matter.

Your RFC wording is exactly what I would propose and would love yes, but this is / could be a huge breaking change for some servers like Navidrome (@deluan)

Hence the more questioning way to gauge servers and clients feedback on that, and possible alternative solutions like adding a parameter to change behavior or whatever.

**Reply by deluan** [reply]
I agree with @spl0k proposal. The only change I'd like to propose is change MUST to SHOULD in 

> If the server doesn't have any art for an entity, the covertArt field MUST be absent

Some servers may implement a lazy image retrieval. They could advertise the `coverArt` field. But this would be ok because if `getCoverArt` is called and the server cannot find the image (in cache, disk or external sources), it would then return a 404 as described. 

This "lazy image retrieval" is how I implemented it, but I've been forced to return a placeholder image when the image does not exist, to stay compliant  with how Subsonic itself implements `getCoverArt`. Returning a 404 in these cases would make sense in this cases, IMO.

**Reply by Tolriq** [reply]
The problem of lazy loading as we discussed on another place, is that apps can't know when to switch to getXXXInfo easily. The data is stored during sync for offline apps, so it would require apps to store both images then fallback. And since we do not expose sizes and everything this also prevent proper decision on what image to display.
Having to call getXXXInfo preventively for all the library is very slow when it's maybe needed for 10%.

Since the need for lazy loading is usually to load the images as getXXXInfo would. Maybe we should add that the servers who always return a data MUST return the image that getXXXInfo would return?

Without sizes exposed this makes the situation hard for client app.
Usually the covertArt is the preferred choice because if the user manually change the image on the server side (if the server have such feature) then it would be returned there.

**Reply by austinried** [reply]
I think it would make more sense to make it about priority. If something has ANY image, one of them (whichever the server defines as the "default" I suppose) should be returned by cover art, so a client can know that if cover art 404s then they don't need to go searching through get info responses for an image.

The client shouldn't care if it's being lazy loaded or even redirected to an external source IMO, that's a server implementation detail.

**Reply by Tolriq** [reply]
Think about offline first clients and how the API currently work.
Using search3 I can get every coverArt for everything in an efficient manner (Since there's currently no other way). If the server returns a covertArt but does not have one I can't know.
So I would need to call thousands of slow getXXXinfo to have alternate urls to fallback just in case? That's highly inefficient.
A client should be able to determine at sync time what image url he will use for a given entity without having to actually get the image and fallback.

A client should not have to rely on error 404 to decide to call another endpoint with data it might not have at that moment to try to gather a fallback image.

So yes it's server implementation details but it should either properly not return a coverArt when it don't have one, or do the fallback itself.

**Reply by austinried** [reply]
Yeah, that's why I'm saying the wording should focus on priority, so that if a client gets a cover art, tries to retrieve it, gets a 404, they can safely assume there are no other images to try in the get info endpoint. That way servers can still lazy load if they need to, and clients don't need to make extra requests for art.

Even for an offline first client, this should be enough, right?

**Reply by Tolriq** [reply]
Yes sorry I missread your message. The question left is that currently a server can have a covertArt representing a low quality tag image and still provide a better one via the getXXXInfo. So this needs to be addressed in some way too at some point. 

But the current pain in the API is really the lack of 404 and returning default images that kinda makes clients ugly :)

**Reply by austinried** [reply]
Thinking about it further though, maybe it's not always possible to determine ahead of time if one of a set of images from potentially multiple sources is "available," so I'm not actually sure if what I said would be reasonable wording.

It'd be easier as a client altogether if all images were just offered up immediately instead of having to call the get info endpoints, (and the other data on there actually...) but I think that has the historic reason behind it of being typically retrieved from thrif party integrations? At least that's what I always thought.

**Reply by Tolriq** [reply]
Initially the coverArt was about the local data from the tags, so servers knew if they had it or not. Then the other APIs where added to get online information.

After that some servers mixed all as there was no official guidance or rule. So coverArt have a different meaning and the issue we currently have.

Not sure it's easy to solve without an extension and change to the API.
